text/0605-eks-rewrite.md

```diff
@@ -0,0 +1,500 @@
+    vpc,
+    vpcSubnets: [{ subnetType: ec2.SubnetType.PRIVATE_ISOLATED }]
+});
+```
+
+### Kubectl Support
+
+You can choose to have CDK create a `Kubectl Handler` - a Python Lambda Function to
+apply k8s manifests using `kubectl apply`. This handler will not be created by default.
+
+To create a `Kubectl Handler`, use `kubectlProviderOptions` when creating the cluster.
+`kubectlLayer` is the only required property in `kubectlProviderOptions`.
+
+```
+new eks.Cluster(this, 'hello-eks', {
+  version: eks.KubernetesVersion.V1_31,
+  # Using this property will create `Kubectl Handler` as custom resource handler
+  kubectlProviderOptions: {
+    kubectlLayer: new KubectlV31Layer(this, 'kubectl'),
+  }
+});
+
+`Kubectl Handler` created along with the cluster will be granted admin permissions to the cluster.
+```
+
+## Permissions and Security
+
+Amazon EKS supports three modes of authentication: CONFIG_MAP, API_AND_CONFIG_MAP, and API.
+`ConfigMap` authentication mode has been deprecated by EKS and the recommended mode is API.
+The new EKS L2 will go a step further and only support API authentication mode.
+All grant functions in EKS will use Access Entry to grant
+permissions to an IAM role/user.
+
+As a result, `AwsAuth` construct in the previous EKS module will not be provided in the new module.
+`grant()` functions are introduced to replace the awsAuth. It’s implemented using
+Access Entry.
+
+Grant Admin Access to an IAM role
+
+```
+cluster.grantAdmin('adminAccess', roleArn, eks.AccessScopeType.CLUSTER);
+```
+
+You can also use general `grantAccess()` to attach a policy to an IAM role/user.
+See https://docs.aws.amazon.com/eks/latest/userguide/access-policies.html for all access policies
+
+```
+# A general grant function is also provided
+cluster.grantAccess('adminAccess', roleArn, [
+  eks.AccessPolicy.fromAccessPolicyName('AmazonEKSClusterAdminPolicy', {
+    accessScopeType: eks.AccessScopeType.CLUSTER,
+  }),
+]);
+```
+
+### Use existing Cluster/Kubectl Handler
+
+This module allows defining Kubernetes resources such as Kubernetes manifests and Helm charts
+on clusters that are not defined as part of your CDK app.
+
+There are 2 scenarios here:
+
+1. Import the cluster without creating a new kubectl Handler
+
+```
+const cluster = eks.Cluster.fromClusterAttributes(this, 'MyCluster', {
+  clusterName: 'my-cluster-name',
+});
+```
+
+This imported cluster is not associated with a `Kubectl Handler`. It means we won't be able to
+invoke `addManifest()` function on the cluster.
+
+To apply a manifest, you need to import the kubectl handler and attach it to the cluster
+
+```
+const kubectlProvider = eks.KubectlProvider.fromKubectlProviderArn(this, 'KubectlProvider', {
+  functionArn: ''
+});
+
+const cluster = eks.Cluster.fromClusterAttributes(this, 'MyCluster', {
+  clusterName: 'my-cluster-name',
+  kubectlProvider: kubectlProvider
+});
+
+cluster.addManifest();
+```
+
+2. Import the cluster and create a new kubectl Handler
```

Comment:
Do we need a separate `fromClusterWithKubectlProvider` function for this use case? 

How about:

```ts
const clusterName = 'my-cluster-name';
const kubectlProvider = new eks.KubectlProvider(this, 'KubectlProvier', {
  clusterName,
})

const cluster = eks.Cluster.fromClusterAttributes(this, 'MyCluster', {
  clusterName,
  kubectlProvider
});
```

This way we just have one `fromXXX` function for cluster import. Do you see any problems with this approach?

Reply:
Are you saying `kubectlProviderOptions` will not be exposed in `fromClusterAttributes()`?

That makes sense to me and I like it. Will update the RFC.